### PR TITLE
Code changes for Probe-level Termination Grace Period Beta

### PR DESCRIPTION
**Pull request by raisaat**
#### What type of PR is this?

/kind feature
/sig node
/priority important-soon

#### What this PR does / why we need it:

Code changes to graduate this KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2238-liveness-probe-grace-period to beta

- [x] Move feature flag to beta, leave as false
- [x] Remove feature gate from kubelet
- [x] Add validation to ensure terminationGracePeriodSeconds is non-negative
  - #98866 
  - https://github.com/kubernetes/kubernetes/pull/103245

#### Which issue(s) this PR fixes:

 KEP issue: https://github.com/kubernetes/enhancements/issues/2238

#### Does this PR introduce a user-facing change?

```release-note
ACTION REQUIRED: Removed the feature flag for probe-level termination grace period from Kubelet. If a user wants to disable this feature on already created pods, they will have to delete and recreate the pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2238-liveness-probe-grace-period
```

### CODE REVIEW

**Comment by ehashman** [comment]
Please make sure you sign the CLA: https://github.com/kubernetes/kubernetes/pull/103168#issuecomment-868018827

File: pkg/api/pod/util_test.go
Comment (ehashman) [inline pkg/api/pod/util_test.go:1231]: This doesn't make sense to me as a case. We're unconditionally evaluating this if the first two cases don't happen, and if it's true, throwing an error. Whether they're equal is a test outcome, not a dev expectation. But, this is also opposite to the default case below, so I think this was just an oversight.

We need a case for `oldPodHasGracePeriod` to assert that the field was dropped. I'd suggest we repurpose this case for that.

Test matrix is

- Feature flag is on:
   - Pod has grace period: unchanged
   - Pod does not have grace period: unchanged
- Feature flag is off:
   - Pod has grace period: should be dropped
   - Pod does not have grace period: unchanged

Thus, I'd expect our cases to be:
- [x] `if enabled || !oldPodHasGracePeriod`: should be unchanged
- [x] Now we know it's disabled. So, we should check next, `if !enabled && newPodHasGracePeriod` and ensure the field is blanked out (note: it is probably clearer to write it the way I've written in this comment rather than trying to be clever with boolean logic)
- [ ] `if !enabled && oldHasGracePeriod`: should be dropped -- need a case for this still

I think we can keep the `default` case below as written unchanged.

**Review by ehashman — COMMENTED** [review]
/sig node
/priority important-soon
/triage accepted

Per https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2238-liveness-probe-grace-period#beta-122 we still need to do the following:

- [ ] Move feature flag to beta, leave as false
- [ ] Remove feature gate from kubelet
- [x] Ensure that when feature gate is off in API server, probe-level TerminationGracePeriodSeconds is blanked out.
- [x] Add validation to ensure terminationGracePeriodSeconds is non-negative. -> see https://github.com/kubernetes/kubernetes/pull/98866

**Comment by raisaat** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

I signed the CLA

File: pkg/api/pod/util_test.go
Comment (raisaat) [inline pkg/api/pod/util_test.go:1211]: `newPod` and `newPodInfo.pod()` should be different after the feature is disabled and the feature is dropped via the `dropDisabledFields` function with the `dropDisabledFields` function written as is right now, right? The reason is that the `dropDisabledFields` function blanks out the `LivenessProbe.TerminationGracePeriodSeconds` and `StartupProbe.TerminationGracePeriodSeconds` fields of the containers in `newPod` while `newPodInfo.pod()` remains unchanged. So `reflect.DeepEqual(newPod, newPodInfo.pod())` should return a `false` for the con

Comment (ehashman) [reply pkg/api/pod/util_test.go:1211]: What you're discussing here should be covered by the `case newPodHasGracePeriod:` below, no?

Comment (raisaat) [reply pkg/api/pod/util_test.go:1206]: If I throw an error directly for `case newPodHasGracePeriod`, 5 test cases fail. So I tried adding `if reflect.DeepEqual(newPod, newPodInfo.pod())` under `case newPodHasGracePeriod` and a `t.Errorf` under that `if` condition. The 5 test cases fail for the latter as well.

File: pkg/api/pod/util_test.go
Comment (raisaat) [inline pkg/api/pod/util_test.go:1200]: left this line and the `for` loop for enabling/disabling the feature (discarded all the other changes I had made previously). The test cases for the feature flag disabled are still failing.

**Comment by raisaat** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

File: pkg/features/kube_features.go
Comment (ehashman) [inline pkg/features/kube_features.go:815]: You may want to add a comment here noting that we are intentionally leaving this as false for Beta right now.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (ehashman) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go:678]: The feature gate should be removed from the kubelet, but we still need to keep the logic inside.

Comment (raisaat) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:678]: Would it be sufficient to comment out this `if` block as it would allow for the feature to remain in the code but not be utilized?

Comment (ehashman) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:678]: No, we shouldn't be leaving comments. The feature gate should be removed.

File: pkg/features/kube_features.go
Comment (ehashman) [inline pkg/features/kube_features.go:815]:
```suggestion
	ProbeTerminationGracePeriod:                    {Default: false, PreRelease: featuregate.Beta}, // Default to false in beta 1.22, set to true in 1.24
```

**Review by ehashman — COMMENTED** [review]
Close, just needs to rebase once the PR with test fixes land I think.

**Comment by ehashman** [comment]
Since https://github.com/kubernetes/kubernetes/pull/103420 merged this should be ready to rebase.

File: pkg/api/pod/util.go
Comment (ehashman) [inline pkg/api/pod/util.go:589]: Will ensure this is blanked out whenever feature flag is disabled: https://github.com/kubernetes/enhancements/pull/2638/files#r632072138

**Review by ehashman — COMMENTED** [review]
/lgtm

**Comment by ehashman** [comment]
/lgtm cancel

Can you please go ahead and remove the unused function per the verify command:

```
pkg/api/pod/util.go:860:6: func probeGracePeriodInUse is unused (U1000)
```

**Review by mrunalp — APPROVED** [review]

**Comment by ehashman** [comment]
/approve

for pkg/features

Need someone to approve the API changes.

**Comment by ehashman** [comment]
/assign @smarterclayton 

You did the alpha API review, can you PTAL?

File: pkg/api/pod/util.go
Comment (liggitt) [reply pkg/api/pod/util.go:589]: why is this being changed... this is non-standard... doesn't this break the following scenario:

1. enable feature
2. create pod with a value in the field
3. disable feature
4. try to make unrelated update to pod (update annotation or something)
5. API server drops value from the field
6. API validation fails because you changed pod spec on update?

Comment (ehashman) [reply pkg/api/pod/util.go:589]: I hadn't thought about this breaking immutability/making it impossible to delete pods, which seems worse than "if you made a pod with the feature on and turn the feature flag off, sorry, it's still on."

I'd say our best path forward here is to keep the second condition as is, add an ACTION REQUIRED release note about the kubelet feature flag removal, and add some documentation that if a user wants to disable this on already created pods, they will have to delete and recreate the pods.

**Comment by ehashman** [comment]
Requested exception: https://groups.google.com/g/kubernetes-sig-node/c/NrqKjVQJrUE/m/NjNSbQJWAAAJ

File: pkg/api/pod/util.go
Comment (liggitt) [reply pkg/api/pod/util.go:589]: That approach matches my expectations on the API side. I'll defer to node reviewers for kubelet feature gate discussion.

**Comment by ehashman** [comment]
To move forward with this:

- Rebase PR on latest master
- Drop the 1st commit that blanks out the field---we don't want to do that
- Add an ACTION REQUIRED release note per https://github.com/kubernetes/kubernetes/pull/103168#discussion_r666571523 about the kubelet feature flag dropping
- I will take a TODO to update the user-facing documentation

**Review by ehashman — COMMENTED** [review]
/lgtm

**Comment by liggitt** [comment]
API bits lgtm

/approve
/hold for #103245 to merge first and for application of milestone

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (aojea) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:678]: if this feature is still beta and you remove the feature gate check, 
what happens if someone wants to disable all beta features or this specific feature?

**Comment by ehashman** [comment]
Looks like this needs a rebase...

**Review by ehashman — COMMENTED** [review]
/lgtm
assuming CI passes

/hold cancel

**Pull request merged**
